README.md

```diff
@@ -1,3 +1,88 @@
+with ZeroMQ.
+
+Querying the process must be done using the bundled python module. This
+can be imported and a function named `query` can be run, e.g. to check
+the health of the service:
+
+```
+>>> import tag_hierarchy_client
+>>> tag_hierarchy_client.query([{'command': 'healthcheck'}], "server")
+```
+
+# Architectural overview
+The following diagram describes which components are bundled with the
+debian package and how they are intended to be used:
+
+![Architecture diagram][docs/hierarchy_daemon.png]
+
+## Communication
+Components communicate via ZeroMQ(ZMQ). We use the patterns REQ-REP
+(request-response) when the python client communicates with the server
+or the statemanager.
+
+From the python perspective, a request has to pass a list of dictionaries
+and expects a list of dictionaries in return. By convention the first
+dictionary of a request constitutes the command. Any number of
+subsequent dictionaries constitute the data associated with the command.
+
+The python module is built in C++ using the package `pybind11`, which
+translates the python dictionaries into C++ native `std::map`. Since
+C++ is statically typed, only a subset of dictionary key and value
+types are permitted, and this is supported on the C++ by using
+`boost::variant`. Currently the supported dictionary type is
+`Dict[str, Union[bool, int, float, none, str, List[int], List[str]]]]`
+Trying to pass objects that do not conform will raise a `TypeError`
+
+## Tag hierarchy service
+The tag hierarchy service keeps a tag hierarchy in memory. To populate
+the tag hierarchy, use the `populate_hierarchy` command. To query the
+hierarchy, the `nodes` command is used.
+
+### Flow
+Populating the hierarchy can take quite a bit of time, therefore the
+package supports running in a statemanager mode which should not be used
+for other purposes than managing the state of the hierarchy.
+
+When the cache is updated, the population process happens against the
+statemanager while the server is still serving requests. When the
+population is done, it is asked to store itself via the backup service.
+When the backup service has stored a new state, it notifies the server
+via ZMQ PUB-SUB which the server will react to and it will get the
+newest state from the backup service and initialize itself with that.
+Finally the statemanager can clear its state so it does not consume
+memory while idle.
+
+## Backup service
+
+### Storing the state
+The backup service is a REQ-REP service. When it gets a message it will
+try to store that message in redis under the key "hierarchy_cache" and
+return `"Success"`. It will also PUBlish a message `"cache_updated"`
+which the server should SUBscribe to and react to by retrieving the
+latest state. 
+
+### Retrieving the state
+If the message sent to the backup service is `"GET_HIERARCHY"` it will
+return a message with the serialized state, which the server can use to
+re-initialize itself with the latest state.
+
+TODO Add a command reference
```

Comment:
todo?

Reply:
I think it is better suited for autogeneration via code commenting. It will be hard to maintain otherwise.

Reply:
ok. then remove todo, and state where to find the autogenerated command reference.